## Problem #155: Counting Capacitor Circuits
[Link to Problem](https://projecteuler.net/problem=155)

### Problem Description

Given $n$ idential capacitors, let $D(n)$ be the number of distinct total capacitance values we can obtain when using up to $n$ equal-valued capacitors.

For example, $D(1) = 1$, $D(2) = 3$, $D(3) = 7$

![For example, using up to 3 capacitors of 60 μF each, we can obtain the following distinct total capacitance values:](./155.png)

### Approach

Lets define what a unit of capacitors is.

A unit $U$ of capacitors is either:

1. a capacitor
2. two sub-units $u_1$ and $u_2$ connected in parallel: $$ C = c_1 + c_2 $$
3. two sub-units $u_1$ and $u_2$ connected in series: $$ \frac{1}{C} = \frac{1}{c_1} + \frac{1}{c_2} $$

Therefore if for a unit $U$ of size $n$ we can find out all the possible values of it's capacitance by:

- iterating how many capacitors does $u_1$ have
- trying each possible capacitance value for $c_1$ and $c_2$

This method is basically **dynamic programming**.

We compute $d(n)$ = the number of unique capacitance values we can obtain when using **exactly** $n$ identical capacitors.

We do this by first computing $C(n)$, the actual list of the capacitances. Then to compute $D(n)$ we just take the cardinal of the union of all $C(n)$, $n \leq 18$

In [1]:
# Fraction class for rational arithmetic
from math import gcd

class Fraction:
    def __init__(self, numerator, denominator):
        self.numerator = numerator
        self.denominator = denominator
        self.simplify()

    def simplify(self):
        common_divisor = gcd(self.numerator, self.denominator)
        self.numerator //= common_divisor
        self.denominator //= common_divisor

    def __add__(self, other):
        new_numerator = self.numerator * other.denominator + other.numerator * self.denominator
        new_denominator = self.denominator * other.denominator
        return Fraction(new_numerator, new_denominator)

    def __mul__(self, other):
        new_numerator = self.numerator * other.numerator
        new_denominator = self.denominator * other.denominator
        return Fraction(new_numerator, new_denominator)

    def __truediv__(self, other):
        new_numerator = self.numerator * other.denominator
        new_denominator = self.denominator * other.numerator
        return Fraction(new_numerator, new_denominator)

    def __str__(self):
        return f"{self.numerator}/{self.denominator}"

    def __eq__(self, other):
        return self.numerator == other.numerator and self.denominator == other.denominator

    def __hash__(self):
        return hash((self.numerator, self.denominator))

In [2]:
# Compute C(n)
limit = 18

C = [set() for _ in range(limit + 1)]

C[1].add(Fraction(1, 1))

for n in range(2, limit + 1):
    for u1 in range(1, n // 2 + 1):
        u2 = n - u1
        for c1 in C[u1]:
            for c2 in C[u2]:
                # Connected in parallel
                C[n].add(c1 + c2)
                # Connected in series
                C[n].add(c1 *  c2 / (c1 + c2))

In [3]:
# Compute D(n)
common = set()

D = [0] * (limit + 1)

for i in range(1, limit + 1):
    common = common.union(C[i])
    D[i] = len(common)

print(D[limit])

3857447


#### Result: **3857447**

### Complexity analysis (optional)

After looking at the values of D(n), it looks like $D(n)$ increases exponentially, so it seems that the time and space complexity of the program is also exponential.

##### Tags: #bruteforce, #dynamic-programming, #exponential 